src/backend/libinput.rs

```diff
@@ -0,0 +1,147 @@
+        self.context.dispatch()?;
+        for event in &mut self.context {
+            match event {
+                Event::Device(device_event) => {
+                    use input::event::device::*;
+                    match device_event {
+                        DeviceEvent::Added(device_added_event) => {
+                            let added = device_added_event.into_event().device();
+
+                            let new_caps = SeatCapabilities {
+                                pointer: added.has_capability(DeviceCapability::Pointer),
+                                keyboard: added.has_capability(DeviceCapability::Keyboard),
+                                touch: added.has_capability(DeviceCapability::Touch),
+                            };
+
+                            let device_seat = added.seat();
+                            self.devices.push(added);
+
+                            let contains = self.seats.contains_key(&device_seat);
+                            if contains {
+                                let old_seat = self.seats.get_mut(&device_seat).unwrap();
+                                {
+                                    let caps = old_seat.capabilities_mut();
+                                    caps.pointer = new_caps.pointer || caps.pointer;
+                                    caps.keyboard = new_caps.keyboard || caps.keyboard;
+                                    caps.touch = new_caps.touch || caps.touch;
+                                }
+                                if let Some(ref mut handler) = self.handler {
+                                    handler.on_seat_changed(old_seat);
+                                }
+                            } else {
+                                let mut hasher = DefaultHasher::default();
+                                device_seat.hash(&mut hasher);
+                                self.seats.insert(device_seat.clone(), Seat::new(hasher.finish(), new_caps));
+                                if let Some(ref mut handler) = self.handler {
+                                    handler.on_seat_created(self.seats.get(&device_seat).unwrap());
+                                }
+                            }
+                        },
+                        DeviceEvent::Removed(device_removed_event) => {
+                            let removed = device_removed_event.into_event().device();
+
+                            // remove device
+                            self.devices.retain(|dev| *dev == removed);
+
+                            let device_seat = removed.seat();
+
+                            // update capabilities, so they appear correctly on `on_seat_changed` and `on_seat_destroyed`.
+                            if let Some(seat) = self.seats.get_mut(&device_seat) {
+                                let caps = seat.capabilities_mut();
+                                caps.pointer = self.devices.iter().any(|x| x.has_capability(DeviceCapability::Pointer));
```

Comment:
I could instead just count the number of devices that expose a certain capability.

Then again devices are grouped into `device_groups` in libinput and I did read in the libinput documentation (can't seem to find it now), that some devices may only expose certain capabilties if the complete group is available. Especially some combined physical devices (e.g. keyboard + touchpad combos) do this.

This means every removed device could change the capabilties of the remaining ones and this is also not as inefficient as it looks, as the amount of devices should not be very high.

I also think this would make the tracking code even uglier, so I would call your idea premature optimization. We should probably consider this at a later stage again and confirm what libinput actually does. (My thinkpad does some of this device_group stuff, so I should be able to test this.)